lib/src/str_util.rs

```diff
@@ -32,105 +32,189 @@ pub enum StringPatternParseError {
-    pub fn is_exact(&self) -> bool {
-        self.as_exact().is_some()
+    pub fn is_case_sensitive_exact(&self) -> bool {
+        self.as_case_sensitive_exact().is_some()
     }
 
     /// Returns a literal pattern if this should match input strings exactly.
     ///
     /// This can be used to optimize map lookup by exact key.
-    pub fn as_exact(&self) -> Option<&str> {
+    pub fn as_case_sensitive_exact(&self) -> Option<&str> {
+        // TODO: Handle trivial case‐insensitive patterns here? It might make people
+        // expect they can use case‐insensitive patterns in contexts where they
+        // generally can’t.
         match self {
-            StringPattern::Exact(literal) => Some(literal),
-            StringPattern::Glob(_) | StringPattern::Substring(_) => None,
+            StringPattern::Exact(literal, CaseSensitivity::Sensitive) => Some(literal),
+            _ => None,
         }
     }
 
     /// Returns the original string of this pattern.
     pub fn as_str(&self) -> &str {
         match self {
-            StringPattern::Exact(literal) => literal,
-            StringPattern::Glob(pattern) => pattern.as_str(),
-            StringPattern::Substring(needle) => needle,
+            StringPattern::Exact(literal, _) => literal,
+            StringPattern::Substring(needle, _) => needle,
+            StringPattern::Glob(pattern, _) => pattern.as_str(),
         }
     }
 
     /// Converts this pattern to a glob string. Returns `None` if the pattern
     /// can't be represented as a glob.
-    pub fn to_glob(&self) -> Option<Cow<'_, str>> {
+    pub fn to_case_sensitive_glob(&self) -> Option<Cow<'_, str>> {
         // TODO: If we add Regex pattern, it will return None.
+        // TODO: Handle trivial case‐insensitive patterns here? It might make people
+        // expect they can use case‐insensitive patterns in contexts where they
+        // generally can’t.
         match self {
-            StringPattern::Exact(literal) => Some(glob::Pattern::escape(literal).into()),
-            StringPattern::Glob(pattern) => Some(pattern.as_str().into()),
-            StringPattern::Substring(needle) if needle.is_empty() => Some("*".into()),
-            StringPattern::Substring(needle) => {
-                Some(format!("*{}*", glob::Pattern::escape(needle)).into())
+            StringPattern::Exact(literal, CaseSensitivity::Sensitive) => {
+                Some(glob::Pattern::escape(literal).into())
+            }
+            StringPattern::Substring(needle, CaseSensitivity::Sensitive) => {
+                if needle.is_empty() {
+                    Some("*".into())
+                } else {
+                    Some(format!("*{}*", glob::Pattern::escape(needle)).into())
+                }
             }
+            StringPattern::Glob(pattern, CaseSensitivity::Sensitive) => {
+                Some(pattern.as_str().into())
+            }
+            _ => None,
```

Comment:
nit: I'll list all variants exhaustively unless that looks too verbose.

Reply:
It’s just the same three variants again but with `CaseSensitivity::Insensitive`. I can change it.